ensure all requests to modules get a response

garlick opened the pull request:
Requests might be "lost" if they arrive while a module is shutting down (e.g. while being unloaded).  This PR ensures those requests get an ENOSYS reply.  A debug message is logged for each message responded to here.

A theory for #570 is that store requests are lost and remain pending in the content-cache when the backing store module is unloaded.

garlick commented:
This isn't ready to merge yet - I'll try rerunning the travis tests a few times to see if I can trigger the bug with this patch applied.

garlick commented:
One failure in two travis runs:
```
ok 7 - version: reports an expected string
# passed all 7 test(s)
1..7
PASS: t0010-generic-utils.t
# ./t0011-content-cache.t: flux session size will be 4
[1456010623.457617] broker.err[0]: content store: Function not implemented
No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself.
The build has been terminated
```

garlick commented:
I am still running down transient travis-ci test failures.  The latest, clearly related to the code this PR is tweaking, is a broker segfault in content-sqlite.t:
```
# ../../t/t0012-content-sqlite.t: flux session size will be 4
[1456164222.744848] broker.err[0]: content store: Function not implemented
flux-start: 0 (pid 49310) Segmentation fault
ok 1 - unload backing store module if loaded
not ok 2 - load content-sqlite module on rank 0
#	
#		flux module load --rank 0 --direct content-sqlite
#	
flux-start: 3 (pid 49318) Killed
flux-start: 2 (pid 49314) Killed
flux-start: 1 (pid 49311) Killed
FAIL: t0012-content-sqlite.t
```

```
Found ./flux-core-0.2.0-25-g1d6968a/_build/t/t0012-content-sqlite.broker.log
[1456164222.534300] broker.debug[0]: content backing store: enabled content-sqlite
[1456164222.536040] broker.info[0]: nodeset: [0-3] (complete)
[1456164222.682049] resource-hwloc.debug[3]: loaded
[1456164222.688346] resource-hwloc.debug[1]: loaded
[1456164222.721117] broker.info[0]: rmmod content-sqlite
[1456164222.721125] content-sqlite.debug[0]: shutdown: begin
[1456164222.721373] broker.debug[0]: content backing store: disabled content-sqlite
[1456164222.744321] content-sqlite.debug[0]: shutdown: 77 entries returned to cache
[1456164222.744352] content-sqlite.debug[0]: responding to post-shutdown content-backing.store
[1456164222.744848] broker.err[0]: content store: Function not implemented
```

grondo commented:
I've found `libSegFault.so` as a handy way to grab a simple backtrace in travis.... I wonder if we should set that up by default in the travis-ci builds...

garlick commented:
Darn, got segfault but no backtrace.  Maybe I got the LD_PRELOAD path wrong.
```
$ export LD_PRELOAD=/lib/x86_64-linux-gnu/libSegFault.so
The command "export LD_PRELOAD=/lib/x86_64-linux-gnu/libSegFault.so" exited with 0.
[big snip]
flux-start: 0 (pid 49268) Segmentation fault
not ok 25 - exercise batching of asynchronous flush to backing store
#	
#	        OLD_COUNT=`flux comms-stats --type int --parse count content` &&
#		flux module remove --rank 0 --direct content-sqlite &&
#		flux module load --rank 0 --direct content-sqlite &&
#		flux module remove --rank 0 --direct content-sqlite &&
#		flux module load --rank 0 --direct content-sqlite &&
#		flux module remove --rank 0 --direct content-sqlite &&
#		flux module load --rank 0 --direct content-sqlite &&
#	        NEW_COUNT=`flux comms-stats --type int --parse count content` &&
#		test $OLD_COUNT -le $NEW_COUNT
#	
# failed 1 among 25 test(s)
1..25
flux-start: 1 (pid 49269) Killed
flux-start: 3 (pid 49274) Killed
flux-start: 2 (pid 49271) Killed
FAIL: t0012-content-sqlite.t
```

grondo commented:
It would be in the `t*.output` file since stderr is redirected there, is that what the output above is from?

Maybe try also setting `SEGFAULT_USE_ALTSTACK` in the environment in case stack is corrupt?

Also you should not need the full path to libSegFault.so, though it doesn't hurt to have full path I guess. (And I did verify you'd have errors from `ld.so` if it couldn't find components of `LD_PRELOAD`)

grondo commented:
Maybe we could also try setting `LIBC_FATAL_STDERR_=1` in the environment which should cause glibc to emit its own fatal error messages to stderr always.

https://stackoverflow.com/questions/32056387/catching-libc-error-messages-redirecting-from-dev-tty?lq=1

grondo commented:
Actually, what happens to stderr of `flux-broker` processes under `flux-start` -- is it still folded onto stderr of `flux-start` process?

garlick commented:
The brokers are started with the subprocess stuff but I think nothing special is done to redirect stderr.  It should just be going to stderr inherited from flux-start if I understand how that works.

Thanks for the suggestions @grondo.  I was able to get another segfault after setting SEGFAULT_USE_ALTSTACK, but still no backtrace in the `t*.output file` (yes that's what's reproduced above).

grondo commented:
> but still no backtrace in the t*.output file

Well, piffle I say! I swear I had this working before. Sorry if I sent you down a rathole!

grondo commented:
Ah, the problem is that `flux-broker` blocks all signals at runtime, thwarting any signal handlers set up by `LD_PRELOAD` magicians.

The following very kludgy patch gets `libSegFault.so` at least basically working in my environment:
(I'm not sure why the broker is trying to handle SEGV, SIGFPE, etc., but perhaps a hack to check for `LD_PRELOAD` or a "debug mode" option for the broker could work here?)

```diff
--- a/src/broker/broker.c
+++ b/src/broker/broker.c
@@ -1446,13 +1446,14 @@ static void broker_block_signals (void)
     sigset_t sigmask;
     sigemptyset(&sigmask);
     sigfillset(&sigmask);
+    sigdelset (&sigmask, SIGSEGV);
     if (sigprocmask (SIG_SETMASK, &sigmask, NULL) < 0)
         err_exit ("sigprocmask");
 }
 
 static void broker_handle_signals (ctx_t *ctx, zlist_t *sigwatchers)
 {
-    int i, sigs[] = { SIGHUP, SIGINT, SIGQUIT, SIGTERM, SIGSEGV, SIGFPE };
+    int i, sigs[] = { SIGHUP, SIGINT, SIGQUIT, SIGTERM, SIGFPE };
     flux_watcher_t *w;
 
     for (i = 0; i < sizeof (sigs) / sizeof (sigs[0]); i++) {
```

garlick commented:
Hah I just noticed that!  Thanks!

garlick commented:
Got one
```
1..7
PASS: t0010-generic-utils.t
# ../../t/t0011-content-cache.t: flux session size will be 4
*** Segmentation fault
Register dump:
 RAX: 0000000000000000   RBX: 0000000000b3f9c0   RCX: 0000000000000002
 RDX: 0000000000b45c00   RSI: 00000000ffffffff   RDI: 0000000000000002
 RBP: 0000000000b047e0   R8 : 00000000ffffffff   R9 : 0000000000000001
 R10: 6c62616c69617661   R11: 0000000000000000   R12: 0000000000000026
 R13: 0000000000afd040   R14: 0000000000b2fb20   R15: 0000000000b2e4d8
 RSP: 00007fffd276b418
 RIP: 00002ab59e0f4231   EFLAGS: 00010287
 CS: 0033   FS: 0000   GS: 0000
 Trap: 0000000e   Error: 00000004   OldMask: 7ffbfaff   CR2: 00000002
 FPUCW: 0000037f   FPUSW: 00000000   TAG: 00000000
 RIP: 00000000   RDP: 00000000
 ST(0) 0000 0000000000000000   ST(1) 0000 0000000000000000
 ST(2) 0000 0000000000000000   ST(3) 0000 0000000000000000
 ST(4) 0000 0000000000000000   ST(5) 0000 0000000000000000
 ST(6) 0000 0000000000000000   ST(7) 0000 0000000000000000
 mxcsr: 1fa0
 XMM0:  00000000000000000000000000000000 XMM1:  00000000000000000000000000000000
 XMM2:  00000000000000000000000000000000 XMM3:  00000000000000000000000000000000
 XMM4:  00000000000000000000000000000000 XMM5:  00000000000000000000000000000000
 XMM6:  00000000000000000000000000000000 XMM7:  00000000000000000000000000000000
 XMM8:  00000000000000000000000000000000 XMM9:  00000000000000000000000000000000
 XMM10: 00000000000000000000000000000000 XMM11: 00000000000000000000000000000000
 XMM12: 00000000000000000000000000000000 XMM13: 00000000000000000000000000000000
 XMM14: 00000000000000000000000000000000 XMM15: 00000000000000000000000000000000
Backtrace:
/lib/x86_64-linux-gnu/libc.so.6(+0x162231)[0x2ab59e0f4231]
/home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/broker/.libs/lt-flux-broker[0x4134cc]
/home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/common/.libs/libflux-core.so.0(+0x14d23)[0x2ab59c668d23]
/home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/common/.libs/libflux-core.so.0(+0xfea4)[0x2ab59c663ea4]
/home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/common/.libs/libflux-core.so.0(+0x35895)[0x2ab59c689895]
/home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/common/.libs/libflux-core.so.0(+0x367e3)[0x2ab59c68a7e3]
/home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/common/.libs/libflux-core.so.0(flux_reactor_run+0x1c)[0x2ab59c661ccc]
/home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/broker/.libs/lt-flux-broker[0x407c4e]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xed)[0x2ab59dfb376d]
/home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/broker/.libs/lt-flux-broker[0x406469]
Memory map:
00400000-00436000 r-xp 00000000 fc:0a 76873043                           /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/broker/.libs/lt-flux-broker
00635000-00636000 r--p 00035000 fc:0a 76873043                           /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/broker/.libs/lt-flux-broker
00636000-00637000 rw-p 00036000 fc:0a 76873043                           /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/broker/.libs/lt-flux-broker
00af5000-00b5c000 rw-p 00000000 00:00 0                                  [heap]
2ab59bf9a000-2ab59bfbc000 r-xp 00000000 fc:0a 92400702                   /lib/x86_64-linux-gnu/ld-2.15.so
2ab59bfbc000-2ab59bfbf000 rw-p 00000000 00:00 0 
2ab59bfcd000-2ab59bfd7000 rw-p 00000000 00:00 0 
2ab59c1bc000-2ab59c1bd000 r--p 00022000 fc:0a 92400702                   /lib/x86_64-linux-gnu/ld-2.15.so
2ab59c1bd000-2ab59c1bf000 rw-p 00023000 fc:0a 92400702                   /lib/x86_64-linux-gnu/ld-2.15.so
2ab59c1bf000-2ab59c1c3000 r-xp 00000000 fc:0a 92400694                   /lib/x86_64-linux-gnu/libSegFault.so
2ab59c1c3000-2ab59c3c2000 ---p 00004000 fc:0a 92400694                   /lib/x86_64-linux-gnu/libSegFault.so
2ab59c3c2000-2ab59c3c3000 r--p 00003000 fc:0a 92400694                   /lib/x86_64-linux-gnu/libSegFault.so
2ab59c3c3000-2ab59c3c4000 rw-p 00004000 fc:0a 92400694                   /lib/x86_64-linux-gnu/libSegFault.so
2ab59c3c4000-2ab59c451000 r-xp 00000000 fc:0a 68355045                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/kvs/.libs/libflux-kvs.so.0.0.0
2ab59c451000-2ab59c650000 ---p 0008d000 fc:0a 68355045                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/kvs/.libs/libflux-kvs.so.0.0.0
2ab59c650000-2ab59c651000 r--p 0008c000 fc:0a 68355045                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/kvs/.libs/libflux-kvs.so.0.0.0
2ab59c651000-2ab59c653000 rw-p 0008d000 fc:0a 68355045                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/kvs/.libs/libflux-kvs.so.0.0.0
2ab59c653000-2ab59c654000 rw-p 00000000 00:00 0 
2ab59c654000-2ab59c6ef000 r-xp 00000000 fc:0a 112578303                  /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/common/.libs/libflux-core.so.0.0.0
2ab59c6ef000-2ab59c8ee000 ---p 0009b000 fc:0a 112578303                  /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/common/.libs/libflux-core.so.0.0.0
2ab59c8ee000-2ab59c8ef000 r--p 0009a000 fc:0a 112578303                  /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/common/.libs/libflux-core.so.0.0.0
2ab59c8ef000-2ab59c8f1000 rw-p 0009b000 fc:0a 112578303                  /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/common/.libs/libflux-core.so.0.0.0
2ab59c8f1000-2ab59c8f2000 rw-p 00000000 00:00 0 
2ab59c8f2000-2ab59c8fa000 r-xp 00000000 fc:0a 112332204                  /home/travis/local/lib/libmunge.so.2.0.0
2ab59c8fa000-2ab59caf9000 ---p 00008000 fc:0a 112332204                  /home/travis/local/lib/libmunge.so.2.0.0
2ab59caf9000-2ab59cafa000 r--p 00007000 fc:0a 112332204                  /home/travis/local/lib/libmunge.so.2.0.0
2ab59cafa000-2ab59cafb000 rw-p 00008000 fc:0a 112332204                  /home/travis/local/lib/libmunge.so.2.0.0
2ab59cafb000-2ab59cb04000 r-xp 00000000 fc:0a 112332218                  /home/travis/local/lib/libjson-c.so.2.0.1
2ab59cb04000-2ab59cd03000 ---p 00009000 fc:0a 112332218                  /home/travis/local/lib/libjson-c.so.2.0.1
2ab59cd03000-2ab59cd04000 r--p 00008000 fc:0a 112332218                  /home/travis/local/lib/libjson-c.so.2.0.1
2ab59cd04000-2ab59cd05000 rw-p 00009000 fc:0a 112332218                  /home/travis/local/lib/libjson-c.so.2.0.1
2ab59cd05000-2ab59cdb0000 r-xp 00000000 fc:0a 112334052                  /home/travis/local/lib/libczmq.so.3.0.0
2ab59cdb0000-2ab59cfaf000 ---p 000ab000 fc:0a 112334052                  /home/travis/local/lib/libczmq.so.3.0.0
2ab59cfaf000-2ab59cfb0000 r--p 000aa000 fc:0a 112334052                  /home/travis/local/lib/libczmq.so.3.0.0
2ab59cfb0000-2ab59cfb2000 rw-p 000ab000 fc:0a 112334052                  /home/travis/local/lib/libczmq.so.3.0.0
2ab59cfb2000-2ab59d004000 r-xp 00000000 fc:0a 112332211                  /home/travis/local/lib/libzmq.so.3.1.0
2ab59d004000-2ab59d203000 ---p 00052000 fc:0a 112332211                  /home/travis/local/lib/libzmq.so.3.1.0
2ab59d203000-2ab59d207000 r--p 00051000 fc:0a 112332211                  /home/travis/local/lib/libzmq.so.3.1.0
2ab59d207000-2ab59d208000 rw-p 00055000 fc:0a 112332211                  /home/travis/local/lib/libzmq.so.3.1.0
2ab59d208000-2ab59d261000 r-xp 00000000 fc:0a 112334069                  /home/travis/local/lib/libsodium.so.17.0.0
2ab59d261000-2ab59d460000 ---p 00059000 fc:0a 112334069                  /home/travis/local/lib/libsodium.so.17.0.0
2ab59d460000-2ab59d461000 r--p 00058000 fc:0a 112334069                  /home/travis/local/lib/libsodium.so.17.0.0
2ab59d461000-2ab59d46a000 rw-p 00059000 fc:0a 112334069                  /home/travis/local/lib/libsodium.so.17.0.0
2ab59d46a000-2ab59d482000 r-xp 00000000 fc:0a 92400699                   /lib/x86_64-linux-gnu/libpthread-2.15.so
2ab59d482000-2ab59d681000 ---p 00018000 fc:0a 92400699                   /lib/x86_64-linux-gnu/libpthread-2.15.so
2ab59d681000-2ab59d682000 r--p 00017000 fc:0a 92400699                   /lib/x86_64-linux-gnu/libpthread-2.15.so
2ab59d682000-2ab59d683000 rw-p 00018000 fc:0a 92400699                   /lib/x86_64-linux-gnu/libpthread-2.15.so
2ab59d683000-2ab59d687000 rw-p 00000000 00:00 0 
2ab59d687000-2ab59d689000 r-xp 00000000 fc:0a 92400695                   /lib/x86_64-linux-gnu/libutil-2.15.so
2ab59d689000-2ab59d888000 ---p 00002000 fc:0a 92400695                   /lib/x86_64-linux-gnu/libutil-2.15.so
2ab59d888000-2ab59d889000 r--p 00001000 fc:0a 92400695                   /lib/x86_64-linux-gnu/libutil-2.15.so
2ab59d889000-2ab59d88a000 rw-p 00002000 fc:0a 92400695                   /lib/x86_64-linux-gnu/libutil-2.15.so
2ab59d88a000-2ab59d88c000 r-xp 00000000 fc:0a 92400602                   /lib/x86_64-linux-gnu/libdl-2.15.so
2ab59d88c000-2ab59da8c000 ---p 00002000 fc:0a 92400602                   /lib/x86_64-linux-gnu/libdl-2.15.so
2ab59da8c000-2ab59da8d000 r--p 00002000 fc:0a 92400602                   /lib/x86_64-linux-gnu/libdl-2.15.so
2ab59da8d000-2ab59da8e000 rw-p 00003000 fc:0a 92400602                   /lib/x86_64-linux-gnu/libdl-2.15.so
2ab59da8e000-2ab59da95000 r-xp 00000000 fc:0a 92400606                   /lib/x86_64-linux-gnu/librt-2.15.so
2ab59da95000-2ab59dc94000 ---p 00007000 fc:0a 92400606                   /lib/x86_64-linux-gnu/librt-2.15.so
2ab59dc94000-2ab59dc95000 r--p 00006000 fc:0a 92400606                   /lib/x86_64-linux-gnu/librt-2.15.so
2ab59dc95000-2ab59dc96000 rw-p 00007000 fc:0a 92400606                   /lib/x86_64-linux-gnu/librt-2.15.so
2ab59dc96000-2ab59dd91000 r-xp 00000000 fc:0a 92400801                   /lib/x86_64-linux-gnu/libm-2.15.so
2ab59dd91000-2ab59df90000 ---p 000fb000 fc:0a 92400801                   /lib/x86_64-linux-gnu/libm-2.15.so
2ab59df90000-2ab59df91000 r--p 000fa000 fc:0a 92400801                   /lib/x86_64-linux-gnu/libm-2.15.so
2ab59df91000-2ab59df92000 rw-p 000fb000 fc:0a 92400801                   /lib/x86_64-linux-gnu/libm-2.15.so
2ab59df92000-2ab59e146000 r-xp 00000000 fc:0a 92400696                   /lib/x86_64-linux-gnu/libc-2.15.so
2ab59e146000-2ab59e345000 ---p 001b4000 fc:0a 92400696                   /lib/x86_64-linux-gnu/libc-2.15.so
2ab59e345000-2ab59e349000 r--p 001b3000 fc:0a 92400696                   /lib/x86_64-linux-gnu/libc-2.15.so
2ab59e349000-2ab59e34b000 rw-p 001b7000 fc:0a 92400696                   /lib/x86_64-linux-gnu/libc-2.15.so
2ab59e34b000-2ab59e350000 rw-p 00000000 00:00 0 
2ab59e350000-2ab59e3a2000 r-xp 00000000 fc:0a 112332209                  /home/travis/local/lib/libsodium.so.13.0.2
2ab59e3a2000-2ab59e5a2000 ---p 00052000 fc:0a 112332209                  /home/travis/local/lib/libsodium.so.13.0.2
2ab59e5a2000-2ab59e5a3000 r--p 00052000 fc:0a 112332209                  /home/travis/local/lib/libsodium.so.13.0.2
2ab59e5a3000-2ab59e5ac000 rw-p 00053000 fc:0a 112332209                  /home/travis/local/lib/libsodium.so.13.0.2
2ab59e5ac000-2ab59e68e000 r-xp 00000000 fc:0a 92275094                   /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.16
2ab59e68e000-2ab59e88d000 ---p 000e2000 fc:0a 92275094                   /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.16
2ab59e88d000-2ab59e895000 r--p 000e1000 fc:0a 92275094                   /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.16
2ab59e895000-2ab59e897000 rw-p 000e9000 fc:0a 92275094                   /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.16
2ab59e897000-2ab59e8ac000 rw-p 00000000 00:00 0 
2ab59e8ac000-2ab59e8c1000 r-xp 00000000 fc:0a 92274941                   /lib/x86_64-linux-gnu/libgcc_s.so.1
2ab59e8c1000-2ab59eac0000 ---p 00015000 fc:0a 92274941                   /lib/x86_64-linux-gnu/libgcc_s.so.1
2ab59eac0000-2ab59eac1000 r--p 00014000 fc:0a 92274941                   /lib/x86_64-linux-gnu/libgcc_s.so.1
2ab59eac1000-2ab59eac2000 rw-p 00015000 fc:0a 92274941                   /lib/x86_64-linux-gnu/libgcc_s.so.1
2ab59eac2000-2ab59eaca000 r-xp 00000000 fc:0a 92400703                   /lib/x86_64-linux-gnu/libnss_compat-2.15.so
2ab59eaca000-2ab59ecc9000 ---p 00008000 fc:0a 92400703                   /lib/x86_64-linux-gnu/libnss_compat-2.15.so
2ab59ecc9000-2ab59ecca000 r--p 00007000 fc:0a 92400703                   /lib/x86_64-linux-gnu/libnss_compat-2.15.so
2ab59ecca000-2ab59eccb000 rw-p 00008000 fc:0a 92400703                   /lib/x86_64-linux-gnu/libnss_compat-2.15.so
2ab59eccb000-2ab59ece2000 r-xp 00000000 fc:0a 92400698                   /lib/x86_64-linux-gnu/libnsl-2.15.so
2ab59ece2000-2ab59eee1000 ---p 00017000 fc:0a 92400698                   /lib/x86_64-linux-gnu/libnsl-2.15.so
2ab59eee1000-2ab59eee2000 r--p 00016000 fc:0a 92400698                   /lib/x86_64-linux-gnu/libnsl-2.15.so
2ab59eee2000-2ab59eee3000 rw-p 00017000 fc:0a 92400698                   /lib/x86_64-linux-gnu/libnsl-2.15.so
2ab59eee3000-2ab59eee5000 rw-p 00000000 00:00 0 
2ab59eee5000-2ab59eeef000 r-xp 00000000 fc:0a 92400604                   /lib/x86_64-linux-gnu/libnss_nis-2.15.so
2ab59eeef000-2ab59f0ef000 ---p 0000a000 fc:0a 92400604                   /lib/x86_64-linux-gnu/libnss_nis-2.15.so
2ab59f0ef000-2ab59f0f0000 r--p 0000a000 fc:0a 92400604                   /lib/x86_64-linux-gnu/libnss_nis-2.15.so
2ab59f0f0000-2ab59f0f1000 rw-p 0000b000 fc:0a 92400604                   /lib/x86_64-linux-gnu/libnss_nis-2.15.so
2ab59f0f1000-2ab59f0fd000 r-xp 00000000 fc:0a 92400800                   /lib/x86_64-linux-gnu/libnss_files-2.15.so
2ab59f0fd000-2ab59f2fc000 ---p 0000c000 fc:0a 92400800                   /lib/x86_64-linux-gnu/libnss_files-2.15.so
2ab59f2fc000-2ab59f2fd000 r--p 0000b000 fc:0a 92400800                   /lib/x86_64-linux-gnu/libnss_files-2.15.so
2ab59f2fd000-2ab59f2fe000 rw-p 0000c000 fc:0a 92400800                   /lib/x86_64-linux-gnu/libnss_files-2.15.so
2ab59f2fe000-2ab59f2ff000 ---p 00000000 00:00 0 
2ab59f2ff000-2ab59f4ff000 rw-p 00000000 00:00 0                          [stack:80418]
2ab59f4ff000-2ab59f500000 ---p 00000000 00:00 0 
2ab59f500000-2ab59f700000 rw-p 00000000 00:00 0                          [stack:80419]
2ab59f700000-2ab59f701000 ---p 00000000 00:00 0 
2ab59f701000-2ab59f901000 rw-p 00000000 00:00 0                          [stack:80420]
2ab59f901000-2ab59f902000 ---p 00000000 00:00 0 
2ab59f902000-2ab59fb02000 rw-p 00000000 00:00 0 
2ab59fb02000-2ab59fb03000 ---p 00000000 00:00 0 
2ab59fb03000-2ab59fd03000 rw-p 00000000 00:00 0                          [stack:80900]
2ab59fd03000-2ab59fd8b000 r-xp 00000000 fc:0a 60031479                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/connector-local/.libs/connector-local.so
2ab59fd8b000-2ab59ff8a000 ---p 00088000 fc:0a 60031479                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/connector-local/.libs/connector-local.so
2ab59ff8a000-2ab59ff8b000 r--p 00087000 fc:0a 60031479                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/connector-local/.libs/connector-local.so
2ab59ff8b000-2ab59ff8d000 rw-p 00088000 fc:0a 60031479                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/connector-local/.libs/connector-local.so
2ab59ff8d000-2ab59ff8e000 rw-p 00000000 00:00 0 
2ab5a0000000-2ab5a0021000 rw-p 00000000 00:00 0 
2ab5a0021000-2ab5a4000000 ---p 00000000 00:00 0 
2ab5a4000000-2ab5a405d000 rw-p 00000000 00:00 0 
2ab5a405d000-2ab5a8000000 ---p 00000000 00:00 0 
2ab5a8000000-2ab5a8043000 rw-p 00000000 00:00 0 
2ab5a8043000-2ab5ac000000 ---p 00000000 00:00 0 
2ab5ac000000-2ab5ac087000 r-xp 00000000 fc:0a 112578314                  /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/modctl/.libs/modctl.so
2ab5ac087000-2ab5ac287000 ---p 00087000 fc:0a 112578314                  /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/modctl/.libs/modctl.so
2ab5ac287000-2ab5ac288000 r--p 00087000 fc:0a 112578314                  /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/modctl/.libs/modctl.so
2ab5ac288000-2ab5ac28a000 rw-p 00088000 fc:0a 112578314                  /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/modctl/.libs/modctl.so
2ab5ac28a000-2ab5ac28b000 rw-p 00000000 00:00 0 
2ab5ac28b000-2ab5ac312000 r-xp 00000000 fc:0a 94693189                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/libmrpc/.libs/libflux-mrpc.so.0.0.0
2ab5ac312000-2ab5ac511000 ---p 00087000 fc:0a 94693189                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/libmrpc/.libs/libflux-mrpc.so.0.0.0
2ab5ac511000-2ab5ac512000 r--p 00086000 fc:0a 94693189                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/libmrpc/.libs/libflux-mrpc.so.0.0.0
2ab5ac512000-2ab5ac514000 rw-p 00087000 fc:0a 94693189                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/libmrpc/.libs/libflux-mrpc.so.0.0.0
2ab5ac514000-2ab5ac515000 rw-p 00000000 00:00 0 
2ab5ac515000-2ab5ac5a2000 r-xp 00000000 fc:0a 68355038                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/kvs/.libs/kvs.so
2ab5ac5a2000-2ab5ac7a1000 ---p 0008d000 fc:0a 68355038                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/kvs/.libs/kvs.so
2ab5ac7a1000-2ab5ac7a2000 r--p 0008c000 fc:0a 68355038                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/kvs/.libs/kvs.so
2ab5ac7a2000-2ab5ac7a4000 rw-p 0008d000 fc:0a 68355038                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/kvs/.libs/kvs.so
2ab5ac7a4000-2ab5ac7a5000 rw-p 00000000 00:00 0 
2ab5ac7a5000-2ab5ac82c000 r-xp 00000000 fc:0a 85408218                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/content-sqlite/.libs/content-sqlite.so
2ab5ac82c000-2ab5aca2b000 ---p 00087000 fc:0a 85408218                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/content-sqlite/.libs/content-sqlite.so
2ab5aca2b000-2ab5aca2c000 r--p 00086000 fc:0a 85408218                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/content-sqlite/.libs/content-sqlite.so
2ab5aca2c000-2ab5aca2e000 rw-p 00087000 fc:0a 85408218                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/content-sqlite/.libs/content-sqlite.so
2ab5aca2e000-2ab5aca2f000 rw-p 00000000 00:00 0 
2ab5aca2f000-2ab5acacd000 r-xp 00000000 fc:0a 92432615                   /usr/lib/x86_64-linux-gnu/libsqlite3.so.0.8.6
2ab5acacd000-2ab5acccd000 ---p 0009e000 fc:0a 92432615                   /usr/lib/x86_64-linux-gnu/libsqlite3.so.0.8.6
2ab5acccd000-2ab5acccf000 r--p 0009e000 fc:0a 92432615                   /usr/lib/x86_64-linux-gnu/libsqlite3.so.0.8.6
2ab5acccf000-2ab5accd1000 rw-p 000a0000 fc:0a 92432615                   /usr/lib/x86_64-linux-gnu/libsqlite3.so.0.8.6
2ab5accd1000-2ab5accd2000 rw-p 00000000 00:00 0 
2ab5accd2000-2ab5acd5b000 r-xp 00000000 fc:0a 118333700                  /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/live/.libs/live.so
2ab5acd5b000-2ab5acf5a000 ---p 00089000 fc:0a 118333700                  /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/live/.libs/live.so
2ab5acf5a000-2ab5acf5b000 r--p 00088000 fc:0a 118333700                  /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/live/.libs/live.so
2ab5acf5b000-2ab5acf5d000 rw-p 00089000 fc:0a 118333700                  /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/live/.libs/live.so
2ab5acf5d000-2ab5acf5e000 rw-p 00000000 00:00 0 
2ab5acf5e000-2ab5acfe4000 r-xp 00000000 fc:0a 126574672                  /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/mecho/.libs/mecho.so
2ab5acfe4000-2ab5ad1e3000 ---p 00086000 fc:0a 126574672                  /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/mecho/.libs/mecho.so
2ab5ad1e3000-2ab5ad1e4000 r--p 00085000 fc:0a 126574672                  /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/mecho/.libs/mecho.so
2ab5ad1e4000-2ab5ad1e6000 rw-p 00086000 fc:0a 126574672                  /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/mecho/.libs/mecho.so
2ab5ad1e6000-2ab5ad1e7000 rw-p 00000000 00:00 0 
2ab5ad1e7000-2ab5ad26d000 r-xp 00000000 fc:0a 10344251                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/wreck/.libs/job.so
2ab5ad26d000-2ab5ad46d000 ---p 00086000 fc:0a 10344251                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/wreck/.libs/job.so
2ab5ad46d000-2ab5ad46e000 r--p 00086000 fc:0a 10344251                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/wreck/.libs/job.so
2ab5ad46e000-2ab5ad470000 rw-p 00087000 fc:0a 10344251                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/wreck/.libs/job.so
2ab5ad470000-2ab5ad471000 rw-p 00000000 00:00 0 
2ab5ad471000-2ab5ad4f7000 r-xp 00000000 fc:0a 10344260                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/wreck/.libs/wrexec.so
2ab5ad4f7000-2ab5ad6f7000 ---p 00086000 fc:0a 10344260                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/wreck/.libs/wrexec.so
2ab5ad6f7000-2ab5ad6f8000 r--p 00086000 fc:0a 10344260                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/wreck/.libs/wrexec.so
2ab5ad6f8000-2ab5ad6fa000 rw-p 00087000 fc:0a 10344260                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/wreck/.libs/wrexec.so
2ab5ad6fa000-2ab5ad6fb000 rw-p 00000000 00:00 0 
2ab5ad6fb000-2ab5ad781000 r-xp 00000000 fc:0a 1037890                    /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/barrier/.libs/barrier.so
2ab5ad781000-2ab5ad981000 ---p 00086000 fc:0a 1037890                    /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/barrier/.libs/barrier.so
2ab5ad981000-2ab5ad982000 r--p 00086000 fc:0a 1037890                    /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/barrier/.libs/barrier.so
2ab5ad982000-2ab5ad984000 rw-p 00087000 fc:0a 1037890                    /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/barrier/.libs/barrier.so
2ab5ad984000-2ab5ad985000 rw-p 00000000 00:00 0 
2ab5ad985000-2ab5ada0c000 r-xp 00000000 fc:0a 35291718                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/resource-hwloc/.libs/resource-hwloc.so
2ab5ada0c000-2ab5adc0c000 ---p 00087000 fc:0a 35291718                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/resource-hwloc/.libs/resource-hwloc.so
2ab5adc0c000-2ab5adc0d000 r--p 00087000 fc:0a 35291718                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/resource-hwloc/.libs/resource-hwloc.so
2ab5adc0d000-2ab5adc0f000 rw-p 00088000 fc:0a 35291718                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/modules/resource-hwloc/.libs/resource-hwloc.so
2ab5adc0f000-2ab5adc10000 rw-p 00000000 00:00 0 
2ab5adc10000-2ab5adc48000 r-xp 00000000 fc:0a 112334057                  /home/travis/local/lib/libhwloc.so.5.6.6
2ab5adc48000-2ab5ade47000 ---p 00038000 fc:0a 112334057                  /home/travis/local/lib/libhwloc.so.5.6.6
2ab5ade47000-2ab5ade48000 r--p 00037000 fc:0a 112334057                  /home/travis/local/lib/libhwloc.so.5.6.6
2ab5ade48000-2ab5ade49000 rw-p 00038000 fc:0a 112334057                  /home/travis/local/lib/libhwloc.so.5.6.6
2ab5ade49000-2ab5adf9b000 r-xp 00000000 fc:0a 92432725                   /usr/lib/x86_64-linux-gnu/libxml2.so.2.7.8
2ab5adf9b000-2ab5ae19a000 ---p 00152000 fc:0a 92432725                   /usr/lib/x86_64-linux-gnu/libxml2.so.2.7.8
2ab5ae19a000-2ab5ae1a2000 r--p 00151000 fc:0a 92432725                   /usr/lib/x86_64-linux-gnu/libxml2.so.2.7.8
2ab5ae1a2000-2ab5ae1a4000 rw-p 00159000 fc:0a 92432725                   /usr/lib/x86_64-linux-gnu/libxml2.so.2.7.8
2ab5ae1a4000-2ab5ae1a5000 rw-p 00000000 00:00 0 
2ab5ae1a5000-2ab5ae1bb000 r-xp 00000000 fc:0a 92275043                   /lib/x86_64-linux-gnu/libz.so.1.2.3.4
2ab5ae1bb000-2ab5ae3ba000 ---p 00016000 fc:0a 92275043                   /lib/x86_64-linux-gnu/libz.so.1.2.3.4
2ab5ae3ba000-2ab5ae3bb000 r--p 00015000 fc:0a 92275043                   /lib/x86_64-linux-gnu/libz.so.1.2.3.4
2ab5ae3bb000-2ab5ae3bc000 rw-p 00016000 fc:0a 92275043                   /lib/x86_64-linux-gnu/libz.so.1.2.3.4
2ab5ae3bc000-2ab5ae3bd000 ---p 00000000 00:00 0 
2ab5ae3bd000-2ab5ae5bd000 rw-p 00000000 00:00 0                          [stack:80901]
2ab5ae5bd000-2ab5ae5be000 ---p 00000000 00:00 0 
2ab5ae5be000-2ab5ae7be000 rw-p 00000000 00:00 0                          [stack:80902]
2ab5ae7be000-2ab5ae7bf000 ---p 00000000 00:00 0 
2ab5ae7bf000-2ab5ae9bf000 rw-p 00000000 00:00 0                          [stack:80903]
2ab5ae9bf000-2ab5ae9c0000 ---p 00000000 00:00 0 
2ab5ae9c0000-2ab5aebc0000 rw-p 00000000 00:00 0                          [stack:80904]
2ab5aebc0000-2ab5aebc1000 ---p 00000000 00:00 0 
2ab5aebc1000-2ab5aedc1000 rw-p 00000000 00:00 0                          [stack:80905]
2ab5aedc1000-2ab5aedc2000 ---p 00000000 00:00 0 
2ab5aedc2000-2ab5aefc2000 rw-p 00000000 00:00 0                          [stack:80906]
2ab5aefc2000-2ab5aefc3000 ---p 00000000 00:00 0 
2ab5aefc3000-2ab5af1c3000 rw-p 00000000 00:00 0                          [stack:80907]
2ab5af1c3000-2ab5af1c4000 ---p 00000000 00:00 0 
2ab5af1c4000-2ab5af3c4000 rw-p 00000000 00:00 0                          [stack:80908]
2ab5af3c4000-2ab5af44a000 r-xp 00000000 fc:0a 94693210                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/connectors/shmem/.libs/shmem.so
2ab5af44a000-2ab5af64a000 ---p 00086000 fc:0a 94693210                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/connectors/shmem/.libs/shmem.so
2ab5af64a000-2ab5af64b000 r--p 00086000 fc:0a 94693210                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/connectors/shmem/.libs/shmem.so
2ab5af64b000-2ab5af64d000 rw-p 00087000 fc:0a 94693210                   /home/travis/build/garlick/flux-core/flux-core-0.2.0-42-ge0c7109/_build/src/connectors/shmem/.libs/shmem.so
2ab5af64d000-2ab5af64e000 rw-p 00000000 00:00 0 
2ab5b0000000-2ab5b0021000 rw-p 00000000 00:00 0 
2ab5b0021000-2ab5b4000000 ---p 00000000 00:00 0 
2ab5b4000000-2ab5b4021000 rw-p 00000000 00:00 0 
2ab5b4021000-2ab5b8000000 ---p 00000000 00:00 0 
2ab5b8000000-2ab5b8021000 rw-p 00000000 00:00 0 
2ab5b8021000-2ab5bc000000 ---p 00000000 00:00 0 
2ab5bc000000-2ab5bc021000 rw-p 00000000 00:00 0 
2ab5bc021000-2ab5c0000000 ---p 00000000 00:00 0 
2ab5c0000000-2ab5c0021000 rw-p 00000000 00:00 0 
2ab5c0021000-2ab5c4000000 ---p 00000000 00:00 0 
2ab5c4000000-2ab5c4021000 rw-p 00000000 00:00 0 
2ab5c4021000-2ab5c8000000 ---p 00000000 00:00 0 
2ab5c8000000-2ab5c8065000 rw-p 00000000 00:00 0 
2ab5c8065000-2ab5cc000000 ---p 00000000 00:00 0 
2ab5cc000000-2ab5cc021000 rw-p 00000000 00:00 0 
2ab5cc021000-2ab5d0000000 ---p 00000000 00:00 0 
2ab5d0000000-2ab5d0031000 rw-p 00000000 00:00 0 
2ab5d0031000-2ab5d4000000 ---p 00000000 00:00 0 
7fffd274d000-7fffd276f000 rw-p 00000000 00:00 0 
7fffd27fe000-7fffd2800000 r-xp 00000000 00:00 0                          [vdso]
ffffffffff600000-ffffffffff601000 r-xp 00000000 00:00 0                  [vsyscall]
flux-start: 0 (pid 45126) Segmentation fault
ok 1 - unload backing store module
flux-start: 1 (pid 45127) Killed
flux-start: 2 (pid 45129) Killed
flux-start: 3 (pid 45132) Killed
```

garlick commented:
OK, I think I found that one.  I've forced a push with a few commits squashed and I'll run travis a few times to be sure this is stable.

I'm inclined to drop 48d2287  "broker: unblock SIGSEGV" and open an issue as that one requires more thought.

@grondo, any harm in leaving in 4815250 "travis-ci: enable SIGSEGV debugging" ?

grondo commented:
> @grondo, any harm in leaving in 4815250 "travis-ci: enable SIGSEGV debugging" ?

I don't think there is any harm and it could be helpful (though it does seem libSegFault.so is having trouble resolving addresses to symbol, file & line).

garlick commented:
Sigh.  The original test from #570 failed:
```
load content-sophia module on rank 0
expecting success: 
        run_timeout 10 flux content flush &&
        NDIRTY=`flux comms-stats --type int --parse dirty content` &&
        test $NDIRTY -eq 0
Alarm clock
not ok 21 - flush rank 0 cache
```
```
[1456248339.828183] broker.debug[0]: content backing store: enabled content-sqlite
[1456248339.828221] broker.debug[0]: content flush begin
[1456248339.828405] broker.debug[0]: content flush +15 (dirty=15 pending=15)
[1456248339.955173] resource-hwloc.debug[2]: loaded
[1456248339.963465] resource-hwloc.debug[0]: loaded
[1456248339.966776] broker.info[0]: nodeset: [0-3] (complete)
[1456248340.054172] resource-hwloc.debug[3]: loaded
[1456248340.059714] resource-hwloc.debug[1]: loaded
[1456248340.097337] broker.info[0]: rmmod content-sqlite
[1456248340.097355] content-sqlite.debug[0]: shutdown: begin
[1456248340.097448] broker.debug[0]: content backing store: disabled content-sqlite
[1456248340.103485] content-sqlite.debug[0]: shutdown: 97 entries returned to cache
[1456248340.103727] broker.info[0]: module content-sqlite exited
[1456248340.150825] broker.info[0]: insmod content-sophia
[1456248350.635639] broker.debug[0]: content backing store: enabled content-sophia
[1456248350.635659] broker.debug[0]: content flush begin
[1456248350.637658] broker.debug[0]: content flush +204 (dirty=204 pending=204)
[1456248351.588389] broker.debug[0]: content flush
[1456248351.638619] broker.debug[0]: content dropcache 207/207
[1456248351.700033] broker.info[0]: rmmod content-sophia
[1456248351.700065] content-sophia.debug[0]: shutdown: begin
[1456248351.700161] broker.debug[0]: content backing store: disabled content-sophia
[1456248351.719158] content-sophia.debug[0]: shutdown: 208 entries returned to cache
[1456248351.726942] broker.info[0]: module content-sophia exited
[1456248351.908272] broker.info[0]: insmod content-sophia
[1456248351.931894] broker.debug[0]: content flush begin
[1456248351.934028] broker.debug[0]: content flush +208 (dirty=208 pending=208)
[1456248371.438674] broker.debug[0]: content backing store: enabled content-sophia
[1456248371.539812] broker.info[0]: shutdown in 0.500s: Exited
[1456248371.539809] content-sophia.debug[0]: broker shutdown in progress
```

grondo commented:
> @grondo, any harm in leaving in 4815250 "travis-ci: enable SIGSEGV debugging" ?

Actually, it appears travis-ci might allow us to set `ulimit -c unlimited` and get coredumps during testing.  This would allow us to traverse `trash*` directories after the checks and analyze any coredumps which would get much higher quality backtraces (and multi thread support).

We could still leave `LD_PRELOAD` for now, but I'll work on this more advanced approach and see if I can get it going.

garlick commented:
> I'll work on this more advanced approach

That sounds pretty great.  No urgency here though as I think I've found the segfault (just not necessarily all the test failures I wanted to fix)

garlick commented:
Just rebased without the broker SIGSEGV unblock patch.

I think this might actually be OK to go in.   There may still be a timing issue with one or both of the  content-sophia "flush rank 0 cache" tests, but the logs for the one failure among many successes did not indicate any of the problems seen before.  It might be a stall in sophia; I've added debug logging so we'll be able to see if that's the case when it happens again.

Since I was rerunning travis tests many times while looking for content related failures, I encountered other intermittent test failures and tried to improve them here too, and also did some gratuitous cleanup such as moving kvs related tests driven by sharness tests to `t/kvs` with the others.   So this got a little de-focused, sorry (I'd be OK with splitting those out if it seems wrong to have them here).

grondo commented:
I'd hate to make more work for you, but I'd feel better about merging if the unrelated tests updates were moved into their own PR. This will also help us when we're reviewing merged PRs to create the NEWS file for the next release.

garlick commented:
No problem!

garlick commented:
The unrelated stuff has been split out to #582 (test cleanup) and #581 (coverity fixes).
What remains is all pretty much related to the problem being solved in this PR.

To summarize: the tests that unload and reload the backing store module would cause messages to be lost and the test to timeout.  Once that was fixed, it exposed a segfault in an error path of the store RPC continuation.  Logging that was added also exposed a logic error in flush batching that was causing the cache to be traversed many more times than necessary.  Finally, content tests were tightened up, and coverage increased to elicit the above problems more frequently (in absence of the fixes).

grondo commented:
Thanks, that is perfect! Any desired ordering to merge your PRs?

grondo commented:
This one seems most important so I'm merging it first. Hope that's ok.

garlick commented:
That's good, and I'll rebase the others once that's done.
